### PR TITLE
ECG Community Detection implementation

### PR DESCRIPTION
**Pull request by ryandewolfe33**
Another community detection algorithm, see #231 

Ensemble Clustering for Graphs (ECG) uses many runs of Louvain (see #488 ) to improve performance and stability.

### CODE REVIEW

File: src/community/ecg.jl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+function ecg(
+    g::AbstractGraph{T};
+    γ=1.0,
+    ensemble_size::Integer=16,
+    min_edge_weight::Real=0.05,
+    distmx::AbstractArray{<:Number}=weights(g),
+    max_moves::Integer=1000,
+    max_merges::Integer=1000,
+    move_tol::Real=10e-10,
+    merge_tol::Real=10e-10,
```
Comment (LoveLow-Global) [inline src/community/ecg.jl]: Just a small note, but could you change the `10e-10` values to `1e-9`? They represent the same number, but the latter is the standard scientific notation.

File: test/community/ecg.jl [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    for g in test_generic_graphs(empty)
+        r = @inferred ecg_weights(g)
+        @test c == r
+    end
+
+    # Undirected loops
+    loops = complete_graph(2)
+    add_edge!(loops, 1, 1)
+    add_edge!(loops, 2, 2)
+    c = sparse([
```
Comment (LoveLow-Global) [inline test/community/ecg.jl:42]: I believe the maximum possible weight is `1.0` for this algorithm, in general. Therefore, I think the outputs being `2.0` is a bug, as the undirected check in `ecg_weights` currently double-counts self-loops.

File: src/community/ecg.jl [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    ensemble_weights = ecg_weights(
+        g;
+        γ=γ,
+        ensemble_size=ensemble_size,
+        distmx=distmx,
+        max_moves=max_moves,
+        move_tol=move_tol,
+        rng=rng,
+    )
+    weights =
```
Comment (LoveLow-Global) [inline src/community/ecg.jl:94]: I noticed an algorithmic detail issue from the weight calculation. According to Equation (1) in the Poulin & Théberge paper (which I found on line 23 of this file), the ensemble co-association adjustment must only be applied to edges within the 2-core of the graph. Any edge outside the 2-core should default to $w_*$.

Currently, if I am not mistaken, it applies the interpolation globally to every edge.
I believe you have to conditionally apply this, using `core_number(g)`.

**Review by LoveLow-Global — CHANGES_REQUESTED** [review]
Thanks for working on this! Adding ECG is a great feature for the community detection.

I've left a few comments. The two main things to check on are ensuring the ensemble weighting only applies to the 2-core of the graph  and a small bug with undirected self-loops.

As this is my first review after being a reviewer here, I may have not given the best review possible. I apologize in advance for this issue, please let me know if there is anything to be improved from my side.

File: src/community/ecg.jl (hunk 1 above)
Comment (ryandewolfe33) [reply src/community/ecg.jl]: Done :)

File: test/community/ecg.jl (hunk 2 above)
Comment (ryandewolfe33) [reply test/community/ecg.jl:42]: Your belief about the edge weights is correct. The diagonal of the adjacency matrix of an undirected graph holds twice the weight of a loop (so that the degree each vertex is the sum of the row/column).

It's only the case for undirected graphs, there is the same test on a directed graph (lines 78-85) where the diagonal is 1.0.

File: src/community/ecg.jl (hunk 3 above)
Comment (ryandewolfe33) [reply src/community/ecg.jl:94]: Thanks for catching this. Without this step, the implementation (and intuition) works nicely for weighted (parallel edges considered as weighted) and directed graphs, but the 2-core is not so obvious in these cases. Also the `core_number(g)` function currently does not support self loops.

My proposed (and now implemented) solution is to have a `min_weight_outside_2core` parameter that calls `core_number(g)` and decreases all edge weights with at least of vertex with core number < 2 to the minimum weight. This way it can be bypassed for any of the edge cases above, and will throw an error if the `core_number` function is not supported for the graph type.

However I'm not super happy with this solution, and would be open to more discussion about the edge cases.

File: test/community/ecg.jl (hunk 2 above)
Comment (LoveLow-Global) [reply test/community/ecg.jl:42]: You are right about the adjacency matrix for undirected graphs preserving the degree sum. My apologies for missing that. Thanks for the reply.

File: src/community/ecg.jl (hunk 3 above)
Comment (LoveLow-Global) [reply src/community/ecg.jl:94]: I think your currect solution is good enough at the moment, I personally cannot think of a better way. I did ask LLMs for any ideas on this, and it did reply as the following.

To keep the API clean, what if we replaced `min_weight_outside_2core` with a simple boolean flag, like `core_weighting::Bool = !is_directed(g)`?
This way:
1. It automatically skips the 2-core logic for directed graphs to avoid ambiguity.
2. For undirected graphs with self-loops (which currently breaks `core_number`), we can either document that `core_weighting=false `must be passed, or handle it internally by passing a loop-free copy of the graph to `core_number` just for the mask.

Please note that this is just a suggestion from a LLM. Your ideas and thoughts have the priority for sure. If you want to move forward with the currect approach, I will do some final checks and merge if there are no other problems.

Comment (ryandewolfe33) [reply src/community/ecg.jl:94]: I think it's clearer to avoid silently apply/not apply the parameter depending on the input, so I'd prefer the current implementation. I've added a bit of documentation and an error message in the function that to better explain the parameter and allowable input combinations.

**Comment by ryandewolfe33** [comment]
Thanks for the helpful review! It's only my ~3rd pull request so I don't feel qualified to say much about your process, but it seems good to me.

File: src/community/ecg.jl
Comment (LoveLow-Global) [inline src/community/ecg.jl]: During my double-check, I noticed the function name `ecg_weights(g; γ, {and so on})` below does not match the `ensemble_weights(g; c, {and so on})` here in the docstring. Can you take a look on this please? Thank you.

**Review by LoveLow-Global — CHANGES_REQUESTED** [review]

**Comment by LoveLow-Global** [comment]
And thank you the kind comments, I really appreciate it! I will try to improve more in the future.

File: src/community/ecg.jl
Comment (ryandewolfe33) [reply src/community/ecg.jl]: Yes thanks for catching. I've updated both doc strings; the style of ecg doc string was inconsistent with the rest of the community detection files.

**Review by LoveLow-Global — APPROVED** [review]

**Comment by LoveLow-Global** [comment]
Thank you so much for the great contribution. I believe it is ready to be merged!

**Comment by LoveLow-Global** [comment]
@Krastanov Hello, I wonder if you can check on the workflow before merging, as well as if it is ready to be merged, as it is my first time reviewing a PR and I want to make it sure. Thank you!

**Comment by Krastanov** [comment]
Thanks both, for the contribution and for the review!

I want to briefly discuss to questions of API design before we merge it (it otherwise seems ready for merge):

- What are the pros and cons of `ensemble_clustering` vs `ecg` as names? `ensemble_clustering` seems a bit more legible to me, especially to someone outside of the field.
- Is the `*_weights` function meant to be public? Do other clustering algorithms implemented here have corresponding `*_weights` functions or something equivalent?

And a question for future work: Given that we now have a few different types of clustering algorithms, does it make sense to start discussing how these can be organized together, e.g. by some general `Clustering` return type and some `clustering(g, ::AlgorithmType)` API?

**Comment by ryandewolfe33** [comment]
Hi Krastanov, thanks for the comments.

- My interpretation is that `ensemble_clustering` is more of a meta-clustering paradigm, whereas `ecg` is the name of this specific ensemble method (with louvain the specific reweighting). I could see a future with a generic ensemble method that uses a different or many internal algorithms.

- I'm not sure, I'd be open to changing it to private. I do know that the weights returned can be quite useful (in layouts for example, see [this notebook](https://github.com/ftheberge/graph-partition-and-measures/blob/master/ECG_extras.ipynb) for some ideas).

-------------------------

I think discussion around a standardized type makes sense, but I'm not sufficiently comfortable with Julia to lead that. I will say that everything right now uses arrays with entries corresponding to community id as the return type, which I think is fine for partitions (every vertex gets exactly one community). Most popular algorithms use partitions (louvain, leiden, infomap, label propagation, some spectral things) so I think it would be fine to standardize around a partition as a result.
 
There are generalizations with outliers (a vertex can have no communities), or overlaps (a vertex can have multiple communities), or fuzzy membership (e.g. 50% in each of two communities) where the return type would have to be different, but at that point it could make sense as it's own package.

I don't have plans for adding any more community detection algorithms right now, so maybe we can move this to an issue where it's more visible and better for a longer discussion.

**Comment by Krastanov** [comment]
Thanks, Ryan! Thanks Jihyung for the review!
